File: ts/src/fastCenterQuantiles.ts
/**
 * Fast algorithm for finding quantiles of pairwise averages (x[i] + x[j])/2.
 *
 * Uses binary search with counting function to find exact quantiles in O(n log(range)) time.
 */

/** Relative epsilon for floating-point comparisons in binary search convergence. */
const RELATIVE_EPSILON = 1e-14;

/**
 * Counts how many pairwise averages (sorted[i] + sorted[j])/2 where i <= j are <= threshold.
 *
 * @param sorted Sorted array of values
 * @param threshold Threshold to count against
 * @returns Number of pairwise averages <= threshold
 */
function countPairsLessOrEqual(sorted: number[], threshold: number): number {
  const n = sorted.length;
  let count = 0;
  // j is not reset: as i increases, threshold decreases monotonically
  let j = n - 1;

  for (let i = 0; i < n; i++) {
    const target = 2 * threshold - sorted[i];
    while (j >= 0 && sorted[j] > target) {
      j--;
    }
    if (j >= i) {
      count += j - i + 1;
    }
  }

  return count;
}

/**
 * Finds the k-th smallest pairwise average using binary search.
 *
 * @param sorted Sorted array of values
 * @param k 1-based rank of the desired quantile
 * @returns The k-th smallest pairwise average
 */
function findExactQuantile(sorted: number[], k: number): number {
  const n = sorted.length;
  const totalPairs = (n * (n + 1)) / 2;

  if (n === 1) return sorted[0];
  if (k === 1) return sorted[0];
  if (k === totalPairs) return sorted[n - 1];

  const min = sorted[0];
  const max = sorted[n - 1];

  // Binary search on value range
  let lo = min;
  let hi = max;

  while (hi - lo > RELATIVE_EPSILON * Math.max(1.0, Math.abs(lo), Math.abs(hi))) {
    const mid = (lo + hi) / 2;
    const count = countPairsLessOrEqual(sorted, mid);
    if (count < k) {
      lo = mid;
    } else {
      hi = mid;
    }
  }

  const target = (lo + hi) / 2;

  // Extract candidates that are close to the target
  const candidates: number[] = [];
  for (let i = 0; i < n; i++) {
    const threshold = 2 * target - sorted[i];

    // Find left boundary using binary search
    let left = i;
    let right = n;
    while (left < right) {
      const m = Math.floor((left + right) / 2);
      if (sorted[m] < threshold - RELATIVE_EPSILON) {
        left = m + 1;
      } else {
        right = m;
      }
    }

    if (
      left < n &&
      left >= i &&
      Math.abs(sorted[left] - threshold) < RELATIVE_EPSILON * Math.max(1.0, Math.abs(threshold))
    ) {
      candidates.push((sorted[i] + sorted[left]) / 2);
    }

    if (left > i) {
      const avgBefore = (sorted[i] + sorted[left - 1]) / 2;
      if (avgBefore <= target + RELATIVE_EPSILON) {
        candidates.push(avgBefore);
      }
    }
  }

  if (candidates.length === 0) {
    return target;
  }

  candidates.sort((a, b) => a - b);

  // Return the candidate that gives exactly k pairs <= it
  for (const c of candidates) {
    if (countPairsLessOrEqual(sorted, c) >= k) {
      return c;
    }
  }

  return target;
}

/**
 * Finds both lower and upper quantile bounds for pairwise averages.
 *
 * @param sorted Sorted array of values
 * @param kLo 1-based rank for lower bound
 * @param kHi 1-based rank for upper bound
 * @returns Tuple of [lower bound, upper bound]
 */
export function fastCenterQuantileBounds(
  sorted: number[],
  kLo: number,
  kHi: number,
): [number, number] {
  const n = sorted.length;
  const totalPairs = (n * (n + 1)) / 2;

  // Clamp ranks to valid range
  const clampedLo = Math.max(1, Math.min(kLo, totalPairs));
  const clampedHi = Math.max(1, Math.min(kHi, totalPairs));

  const lower = findExactQuantile(sorted, clampedLo);
  const upper = findExactQuantile(sorted, clampedHi);

  // Ensure lower <= upper
  return lower > upper ? [upper, lower] : [lower, upper];
}

File: go/fast_center_quantiles.go
package pragmastat

import (
	"math"
	"sort"
)

// relativeEpsilon is the tolerance for floating-point comparisons in binary search convergence.
const relativeEpsilon = 1e-14

// fastCenterQuantileBounds computes both lower and upper bounds from pairwise averages.
// Uses binary search with counting function to avoid materializing all N(N+1)/2 pairs.
func fastCenterQuantileBounds(sorted []float64, marginLo, marginHi int64) (lo, hi float64) {
	n := len(sorted)
	totalPairs := int64(n) * int64(n+1) / 2

	if marginLo < 1 {
		marginLo = 1
	}
	if marginLo > totalPairs {
		marginLo = totalPairs
	}
	if marginHi < 1 {
		marginHi = 1
	}
	if marginHi > totalPairs {
		marginHi = totalPairs
	}

	lo = fastCenterFindExactQuantile(sorted, marginLo)
	hi = fastCenterFindExactQuantile(sorted, marginHi)

	if lo > hi {
		lo, hi = hi, lo
	}
	return lo, hi
}

// fastCenterCountPairsLessOrEqual counts pairwise averages <= target value.
// Uses O(n) two-pointer algorithm.
func fastCenterCountPairsLessOrEqual(sorted []float64, target float64) int64 {
	n := len(sorted)
	var count int64
	// j is not reset: as i increases, threshold decreases monotonically
	j := n - 1

	for i := 0; i < n; i++ {
		threshold := 2*target - sorted[i]

		for j >= 0 && sorted[j] > threshold {
			j--
		}

		if j >= i {
			count += int64(j - i + 1)
		}
	}

	return count
}

// fastCenterFindExactQuantile finds the exact k-th pairwise average using selection algorithm.
func fastCenterFindExactQuantile(sorted []float64, k int64) float64 {
	n := len(sorted)
	totalPairs := int64(n) * int64(n+1) / 2

	if n == 1 {
		return sorted[0]
	}

	if k == 1 {
		return sorted[0]
	}

	if k == totalPairs {
		return sorted[n-1]
	}

	lo := sorted[0]
	hi := sorted[n-1]
	const eps = relativeEpsilon

	for hi-lo > eps*math.Max(1.0, math.Max(math.Abs(lo), math.Abs(hi))) {
		mid := (lo + hi) / 2
		countLessOrEqual := fastCenterCountPairsLessOrEqual(sorted, mid)

		if countLessOrEqual >= k {
			hi = mid
		} else {
			lo = mid
		}
	}

	target := (lo + hi) / 2
	var candidates []float64

	for i := 0; i < n; i++ {
		threshold := 2*target - sorted[i]

		left := i
		right := n

		for left < right {
			m := (left + right) / 2
			if sorted[m] < threshold-eps {
				left = m + 1
			} else {
				right = m
			}
		}

		if left < n && left >= i && math.Abs(sorted[left]-threshold) < eps*math.Max(1.0, math.Abs(threshold)) {
			candidates = append(candidates, (sorted[i]+sorted[left])/2)
		}

		if left > i {
			avgBefore := (sorted[i] + sorted[left-1]) / 2
			if avgBefore <= target+eps {
				candidates = append(candidates, avgBefore)
			}
		}
	}

	if len(candidates) == 0 {
		return target
	}

	// Sort candidates
	sort.Float64s(candidates)

	for _, candidate := range candidates {
		countAtCandidate := fastCenterCountPairsLessOrEqual(sorted, candidate)
		if countAtCandidate >= k {
			return candidate
		}
	}

	return target
}
